src/hotspot/share/gc/z/zMarkStackAllocator.cpp

```diff
@@ -55,12 +55,74 @@ ZMarkStackSpace::ZMarkStackSpace() :
 
   // Register mark stack space start
   ZMarkStackSpaceStart = _start;
+
+  // Prime space
+  _end += expand_space();
 }
 
 bool ZMarkStackSpace::is_initialized() const {
   return _start != 0;
 }
 
+size_t ZMarkStackSpace::size() const {
+  return _end - _start;
+}
+
+static bool is_high_usage(size_t size) {
+  // Consider usage to be high if we've used more than a 8th of the available space.
+  // The available space (controlled by ZMarkStackSpaceLimit) is by default 8G, so
+  // the high usage threshold will by default be 1G. The vast majority of workloads
+  // will use a few hundred megabytes of space at most, so hitting the high usage
+  // limit should be a rare event. At the same time, workloads that do see excessive
+  // mark stack usage will typically see continuous mark stack space growth until
+  // the high usage limit it hit, so setting the high usage limit too high will in
+  // those situations just waste memory.
+  return size > (ZMarkStackSpaceLimit / 8);
+}
+
+size_t ZMarkStackSpace::expand_space() {
+  const size_t expand_size = ZMarkStackSpaceExpandSize;
+  const size_t old_size = size();
+  const size_t new_size = old_size + expand_size;
+
+  if (new_size > ZMarkStackSpaceLimit) {
+    // Expansion limit reached. This is a fatal error since we
+    // currently can't recover from running out of mark stack space.
+    fatal("Mark stack space exhausted. Use -XX:ZMarkStackSpaceLimit=<size> to increase the "
+          "maximum number of bytes allocated for mark stacks. Current limit is " SIZE_FORMAT "M.",
+          ZMarkStackSpaceLimit / M);
+  }
+
+  if (!_high_usage && is_high_usage(new_size)) {
+    Atomic::store(&_high_usage, true);
+  }
+
+  log_debug(gc, marking)("Expanding mark stack space: " SIZE_FORMAT "M->" SIZE_FORMAT "M (%s Usage)",
+                         old_size / M, new_size / M, _high_usage ? "High" : "Low");
```

Comment:
A bit strange to only have "Low" and "High" usage, maybe "Normal" and "High" would be more accurate.

Reply:
Fixed!